coveo-functools/coveo_functools/flex.py

```diff
@@ -0,0 +1,300 @@
+from __future__ import annotations
+
+import functools
+import inspect
+from typing import (
+    Type,
+    TypeVar,
+    Optional,
+    Any,
+    get_args,
+    get_origin,
+    Union,
+    Dict,
+    Callable,
+    Tuple,
+    overload,
+    Final,
+    List,
+)
+
+from coveo_functools.annotations import find_annotations
+from coveo_functools.casing import unflex, flexcase
+from coveo_functools.dispatch import dispatch
+from coveo_functools.exceptions import UnsupportedAnnotation
+
+_ = unflex, flexcase  # mark them as used (forward compatibility vs docs)
+
+
+JSON_TYPES = (
+    str,
+    bool,
+    int,
+    float,
+    type(None),
+    dict,
+)  # list omitted to support list of custom types
+PASSTHROUGH_TYPES = {None, Any, *JSON_TYPES}
+
+T = TypeVar("T")
+
+TypeHint = Any  # :shrug:
+
+RealClass = Type[T]
+RealFunction = Callable[..., T]
+RealObject = Union[RealClass, RealFunction]
+WrappedClass = Type[T]
+WrappedFunction = Callable[..., T]
+WrappedObject = Union[RealClass, RealFunction]

```

Comment:
Are you sure that what you want here is not
```suggestion
WrappedObject = Union[WrappedClass, WrappedFunction]

```
?

Reply:
hey good catch!